### PR TITLE
Optimize System.DateTime

### PR DESCRIPTION
**Pull request by pentp**
Makes DateTime.UtcNow about 5% faster (over 90% time is spent in OS calls now).
Also changed most of DateTime calculations to use unsigned integers.

### CODE REVIEW

**Comment by jkotas** [comment]
Related to https://github.com/dotnet/runtime/pull/44771

File: src/coreclr/src/System.Private.CoreLib/src/System/DateTime.Windows.CoreCLR.cs
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/DateTime.Windows.CoreCLR.cs:14]: It may be nice to delete these and switch to fully managed implementation as done in the other PR to avoid the need for this dual-maintenance.

Comment (pentp) [reply src/coreclr/src/System.Private.CoreLib/src/System/DateTime.Windows.CoreCLR.cs:14]: But the other PR is going to supersede this part anyway? Or is that still far off?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1476]: I do not think it is worth optimizing these using ThrowHelper.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1127]: This is just multiplying two uints. Why does this need to use hardware intrinsics?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1012]: This needs mention in 3rd party notices.

Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1012]: This is using just a number, not the any code?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1127]: The default multiplication by reciprocal needs to account for a 32-bit uint, but by knowing that he input is limited to 14 bits (9999), it's possible to achieve it with a single multiplication (minus all the shifts/subs). Using MULX here to get the high 32-bits of the multiplication.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1476]: This isn't an optimization but just reducing code size/duplication.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1012]: We prefer to err on giving more credit than necessary. Also, it makes the IP scans easier.

We do have https://github.com/lemire/fastmod referenced in the 3rd party notices already. Does it generate the same result?

Also, isn't the magic div/mod already handled by the JIT?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1012]: https://github.com/dotnet/runtime/blob/703145649baf152e3678825be1849f6fd1bd2b24/src/coreclr/src/jit/lower.cpp#L5067

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1127]: Ah, ok. Is this tricky code really adding significant benefits over what the JIT does for you already?

**Comment by tarekgh** [comment]
@pentp I am off properly will not be able to go deeper in this review. I am seeing you are touching the leap seconds handling in the code and I am not sure if the changes would affect that. did you test with a leap seconds on in any system?

**Comment by tarekgh** [comment]
@jkotas would be good to not merge this till it is carefully reviewed and tested with leap seconds.

**Comment by jkotas** [comment]
> @jkotas would be good to not merge this till it is carefully reviewed and tested with leap seconds.

Agree. I won't merge this before you and @GrabYourPitchforks sign off on this.

**Comment by ViktorHofer** [comment]
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: isn't `(uint)(year * -1030792151) > 171798691)` more efficient than this BMI2 code?

**Comment by jkotas** [comment]
@pentp I have deleted the DateTime FCalls in #46690 that will make this delta smaller. Could you please resolve the conflict?

**Comment by tarekgh** [comment]
Hi @pentp, do you have a chance to resolve the merge conflict?

**Comment by pentp** [comment]
Yes, I'll rebase and fix the conflicts.
I've thought about how to write tests for the leap second paths also and now that the native part has been removed it should be possible to use function pointers in checked builds for `TzSpecificLocalTimeToSystemTime/SystemTimeToFileTime` (like `s_pfnGetSystemTimeAsFileTime`) and then write tests that intercept these calls so that `IsValidTimeWithLeapSeconds` can return a predetermined value. Alternatively, the entire `IsValidTimeWithLeapSeconds` call could be intercepted (perf less important as it should be a rare call). Does that sound like something worth investigating?

**Comment by tarekgh** [comment]
@pentp 

> Yes, I'll rebase and fix the conflicts.

Thanks a lot.

> I've thought about how to write tests for the leap second paths also and now that the native part has been removed it should be possible to use function pointers in checked builds for TzSpecificLocalTimeToSystemTime/SystemTimeToFileTime (like s_pfnGetSystemTimeAsFileTime) and then write tests that intercept these calls so that IsValidTimeWithLeapSeconds can return a predetermined value. Alternatively, the entire IsValidTimeWithLeapSeconds call could be intercepted (perf less important as it should be a rare call). Does that sound like something worth investigating?

I didn't closely review your original changes so I am not clear why need to intercept the calls to IsValidTimeWithLeapSeconds? is it part of the optimization we are trying to do here? or this is just to do some validation? If this is not related to the optimization, then I would say let's focus to get the optimization first and have some perf numbers how much we are gaining.

**Comment by tarekgh** [comment]
@pentp sorry to ping you again. Would you be able to get this soon? I am asking because we are trying to close long opened PR's. Thanks!

**Comment by pentp** [comment]
Rebased and resolved conflicts. Test failures seem to be #47374.

**Comment by tarekgh** [comment]
Thanks @pentp. I'll take a look soon.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:167]: >private [](start = 8, length = 7)

nit: why didn't use same syntax of => here?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:201]: >WithLeapSecond [](start = 16, length = 14)

why can't just do 
```C#
_dateData = DateToTicks(year, month, day) + TimeToTicks(hour, minute, second);
```

and then call VlidateLeapSeconds method on `this` after that? the current pattern looks a little bit strange to me.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:223]: >WithLeapSecond [](start = 16, length = 14)

same as last comment

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:248]: >WithLeapSecond [](start = 16, length = 14)

ditto

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:421]: >ThrowOutOfRange [](start = 24, length = 15)

don't we have any other helper method doing the same instead of defining this one here?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:735]: >0 [](start = 28, length = 1)

why you didn't use MinTicks here?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:796]: > return new DateTime(universalTicks | KindUtc); [](start = 11, length = 47)

wouldn't this can cause throwing exception in the constructor? I think KindUtc value is greater than MaxTicks

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:827]: >return new DateTime(value.UTicks | ((ulong)kind << KindShift)) [](start = 12, length = 62)

this also can cause the constructor to throw when checking against the MaxTicks?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:832]: >LocalMask [](start = 29, length = 9)

could you please add a comment here telling we are checking for Local Kind?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:827]: ignore this. I noticed the constructor that takes ulong. 

---
In reply to: [565693181](https://github.com/dotnet/runtime/pull/45479#discussion_r565693181) [](ancestors = 565693181)

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:796]: ignore this. I noticed the constructor that takes ulong. 

---
In reply to: [565691515](https://github.com/dotnet/runtime/pull/45479#discussion_r565691515) [](ancestors = 565691515)

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1265]: >LocalMask [](start = 29, length = 9)

please add comment we are testing for local kind or convert this check to inlined property.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: It felt a bit long for a one liner and I saw that mostly such methods are written in the long form. I'm generally in favor of the arrow syntax (with the arrow starting on the next line for longer lines).

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: This was to save some prolog overhead for the common case, but it turns out to be a quite small difference, so removed these wrapper methods.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:381]: There's only one other identical use for this in `Calendar` (+1 with different parameter which doesn't currently have a `ExceptionArgument` enum value) - should I move it to `ThrowHelper`?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: Removed the `LocalMask` constant entirely and switched to using `KindLocal` everywhere - this should help with clarity.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:381]: I prefer moving it to the helper but would be ok you think otherwise too.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs:29]: >tmp [](start = 30, length = 3)

what is the benefit having this tmp?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:381]: In light of https://github.com/dotnet/runtime/issues/46940#issuecomment-759923313 I'd leave it just here for now.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:381]: that is fine. Thanks!

File: src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs:29]: This avoids a duplicate memory access for the remainder calculation.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs:29]: Cool. do you mind add a little comment on that to help the code readers understand it?

**Comment by tarekgh** [comment]
@pentp thanks a lot. The changes LGTM. As it looks this changes will give us some perf gain. do we have some perf numbers for that. (at least in the basic scenarios with DateTime)?

**Review by tarekgh — APPROVED** [review]

**Comment by pentp** [comment]
For `DateTime.UtcNow` it's about 6% faster (x64 asm size from 625 to 413), `DateTime.Now` is about 10% faster, `TryParseExact` for round-trip format `O` is about 4% faster.
The other methods I haven't measured properly, but they all have reduced asm code size.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (mj1856) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:361]: Why create `UTicks` instead of just making `InternalTicks ` a `ulong`?

Also I'm not sure why `InternalTicks` is `internal` anyway - it doesn't appear used outside this file.  Shouldn't it be `private`?

**Comment by mj1856** [comment]
I got through about half of this so far, and will review further tomorrow.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: Having it as `59.9999999` is the correct solution to avoid the out-of-order problem.

Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: I removed the `WithLeapSecond` functions already, but I'm not sure what's the most correct solution for the 59.nnnxx issue - mapping 60.500 to 59.99999 sounds fine (better than 59.5 and basically the same as 59.999), but what about the constructors that are accurate to a second only (mapping 60 to 59.99999 might look a bit weird/unexpected)? Or the ones using a custom calendar?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: We need to set `59.9999999` when the call is coming from the UtcNow/Now only. For the constructors, we shouldn't change anything from the passed parameters. i.e. if someone passed second 60. Should be treated as 59 without adding any milliseconds or any other ticks. such constructors is not caring about the milliseconds/ticks accuracy anyway.

**Comment by tarekgh** [comment]
@pentp did you have time to address the remaining comments? I think we should be good to go after that.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:360]: Don't know why there was `InternalTicks` at all - it was exactly the same as `Ticks`. Removed it and replaced all uses with `Ticks`.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:360]: That is make sense.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:255]: This part need to be fixed. pass the input millisecond instead of 999

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:275]: same here, need to pass the input milliseconds instead of 999

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:295]: ditto. pass the input milliseconds. 

I'll stop commenting on that as we need to fix that in all instances. We need only to change the milliseconds and extra ticks in case of `UtcNow` only.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (pentp) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:255]: Converting 60.nnn into 59.nnn sounds like the worst possible option (others being 59.999 or 59.9999999). From the original code structure it looked like an oversight not an intentional decision?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: > Converting 60.nnn into 59.nnn sounds like the worst possible option (others being 59.999 or 59.9999999). From the original code structure it looked like an oversight not an intentional decision?

You never know what is the intentional of the constructor caller. At least we are sure this is not coming from `UtcNow` as we handle that there. We change the second to 59 because we had no other options and it is easy to explain. Changing the milliseconds can be confusing especially if this is not desired by the caller.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: Thanks for addressing the feedback!

**Comment by tarekgh** [comment]
@pentp thanks for getting this done. this is amazing.

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/issues/45318

**Pull request merged**
